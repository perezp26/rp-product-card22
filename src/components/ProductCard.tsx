import styles from '../styles/styles.module.css'
import { useProduct } from '../hooks/useProduct';
import React, { createContext,  CSSProperties } from 'react';

import { ProductContextProps, Product, onChngeArgs, InitialValues, ProductCardHandlers } from '../interfaces/interfaces';

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;



export interface Props {
    product : Product,
    //children?: ReactElement | ReactElement[],
    children: ( args : ProductCardHandlers ) => JSX.Element 
    className?: string, 
    style?: CSSProperties,
    onChange?: (args : onChngeArgs ) => void,
    value?: number,
    initialValues?: InitialValues,
}

export const ProductCard = ({ product, children, className, style, onChange, value, initialValues }: Props) => {

    const {counter, increaseBy, maxCount, isMaxCountReached, reset} = useProduct( { onChange, product, value, initialValues } );

  return (
        <Provider value={{
            product,
            counter,
            increaseBy,
            maxCount
        }}>
             <div 
                className={ `${styles.productCard} ${ className }` } 
                style={ style }
            >
                
                { 
                    children( {
                        count : counter, 
                        isMaxCountReached,
                        maxCount :  initialValues?.maxCount,
                        product,
                        increaseBy,
                        reset
                        
                    }) 
                }
                
            </div>
        </Provider>
           
  )
}
