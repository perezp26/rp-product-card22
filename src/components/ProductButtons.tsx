import React, { CSSProperties, useCallback, useContext } from 'react';

import styles from '../styles/styles.module.css'
import { ProductContext } from './ProductCard';

interface Props {
    className?: string;
    style?: CSSProperties 
}

export const ProductsButtons = ({ className, style } : Props )  =>{


    const {increaseBy, counter, maxCount } = useContext(ProductContext);

    const isMaxReached = useCallback(
      () => !!maxCount && counter === maxCount,
      [counter, maxCount],
    )
    
    //TODO: isMaxReached = useCollback, [count, maxCount] 
    // true si el coutn == maxCount
    // false si no lo es

    return(
        <div 
            className={ `${styles.buttonsContainer} ${ className }`}
            style = { style }
        >
                    <button 
                        className={ styles.buttonMinus }
                        onClick = { () => increaseBy(-1) }> - </button>
                    <div className={ styles.countLabel }> { counter } </div>
                    <button 
                        className={ `${styles.buttonAdd} ${ isMaxReached() && styles.disable}` }
                        onClick = { () => increaseBy(1) }> + </button>
        </div>
    )
}