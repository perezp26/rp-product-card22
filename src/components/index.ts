import { ProductCard  as ProductCardHOC } from './ProductCard';
import { ProductCardHOCProps } from '../interfaces/interfaces';

import { ProductTitle } from './ProductTitle';
import { ProductImage } from './ProductImage';
import { ProductsButtons } from './ProductButtons';

export { ProductTitle } from './ProductTitle';
export { ProductImage } from './ProductImage';
export { ProductsButtons } from './ProductButtons';

export const ProductCard : ProductCardHOCProps = Object.assign( ProductCardHOC, {
    Title: ProductTitle,
    Image: ProductImage,
    Buttons: ProductsButtons
});


export default ProductCard;