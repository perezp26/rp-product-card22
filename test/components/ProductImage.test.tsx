import React from 'react';
import render from 'react-test-renderer';
import { ProductCard } from '../../src/components/ProductCard';
import { ProductImage } from '../../src/components/ProductImage';
import { product2 } from './data/products';


describe('ProductImge', () => {

    test('debe de mostar el componente correctamente con la imagen personalizado', () =>{

        const wrapper = render.create(
            <ProductImage img='' />
        )

        expect( wrapper.toJSON() ).toMatchSnapshot();

    })

    test('should debe de mostar el componente con la imagen del producto ', () => { 

        const wrapper = render.create(
           <ProductCard product={ product2 } >
               {
                   () => (
                       <ProductImage />
                   )
               }
           </ProductCard>
        )

        expect( wrapper.toJSON() ).toMatchSnapshot();

     })
  
});