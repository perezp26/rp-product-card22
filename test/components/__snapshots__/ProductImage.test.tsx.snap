// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImge debe de mostar el componente correctamente con la imagen personalizado 1`] = `
<img
  alt="Product"
  className="productImg undefined"
  src="https://hola.jpg"
/>
`;

exports[`ProductImge should debe de mostar el componente con la imagen del producto  1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Product"
    className="productImg undefined"
    src="./coffee-mug.png"
  />
</div>
`;
